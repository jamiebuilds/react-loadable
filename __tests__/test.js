'use strict';

const React = require('react');
const renderer = require('react-test-renderer');
const Loadable = require('../src');

function waitFor(delay) {
  return new Promise(resolve => {
    setTimeout(resolve, delay);
  });
}

function createLoader(delay, loader, error) {
  return () => {
    return waitFor(delay).then(() => {
      if (loader) {
        return loader();
      } else {
        throw error;
      }
    });
  };
}

function MyLoadingComponent(props) {
  return <div>MyLoadingComponent {JSON.stringify(props)}</div>;
}

function MyComponent(props) {
  return <div>MyComponent {JSON.stringify(props)}</div>;
}

function CodeSplitRenderer({ codeSplit, ...props }) {
  // Enable component to render with or without a 'loading' component configured
  const { isLoading, loaded } = typeof codeSplit.isLoading === 'boolean' ? codeSplit : { loaded: codeSplit };
  if (isLoading) {
    return <MyLoadingComponent {...codeSplit} />;
  }

  return <loaded.MyComponent {...props}/>;
}

function CodeSplitMapRenderer({ codeSplit, ...props }) {
  const { loaded } = codeSplit;
  return whenLoaded(codeSplit, () => (
    <div>
      <loaded.a.MyComponent {...props}/>
      <loaded.b.MyComponent {...props}/>
    </div>
  ));
}

function whenLoaded(state, loadedComponent) {
  const { isLoading, error, loaded } = state;
  if (isLoading || error) {
    return <div>MyLoadingComponent {JSON.stringify(state)}</div>;
  }

  if (loaded) {
    return loadedComponent();
  }

  return null;
}

afterEach(async () => {
  try {
    await Loadable.preloadAll();
  } catch (err) {}
});

test('loading success', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => MyComponent),
    loading: MyLoadingComponent
  });

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

  let component2 = renderer.create(<LoadableMyComponent prop="bar" />);

  expect(component2.toJSON()).toMatchSnapshot(); // reload
});

test('delay and timeout', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(300, () => MyComponent),
    loading: MyLoadingComponent,
    delay: 100,
    timeout: 200,
  });

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(100);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(100);
  expect(component1.toJSON()).toMatchSnapshot(); // timed out
  await waitFor(100);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded
});

test('loading error', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, null, new Error('test error')),
    loading: MyLoadingComponent
  });

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // errored
});

test('server side rendering', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => require('../__fixtures__/component')),
    loading: MyLoadingComponent,
  });

  await Loadable.preloadAll();

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // serverside
});

test('server side rendering es6', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => require('../__fixtures__/component.es6')),
    loading: MyLoadingComponent,
  });

  await Loadable.preloadAll();

  let component = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component.toJSON()).toMatchSnapshot(); // serverside
});

test('preload', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => MyComponent),
    loading: MyLoadingComponent
  });

  let promise = LoadableMyComponent.preload();
  await waitFor(200);

  let component1 = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component1.toJSON()).toMatchSnapshot(); // still loading...
  await promise;
  expect(component1.toJSON()).toMatchSnapshot(); // success

  let component2 = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component2.toJSON()).toMatchSnapshot(); // success
});

test('render', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => ({ MyComponent })),
    loading: MyLoadingComponent,
    render(loaded, props) {
      return <loaded.MyComponent {...props}/>;
    }
  });
  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // success
});

test('render element', async () => {
    let LoadableMyComponent = Loadable({
        loader: createLoader(400, () => ({ MyComponent })),
        loading: MyLoadingComponent,
        render: <CodeSplitRenderer />
    });
    let component = renderer.create(<LoadableMyComponent prop="baz" />);
    expect(component.toJSON()).toMatchSnapshot(); // initial
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // loading
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // success
});

test('render without loading prop', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => ({ MyComponent })),
    render(state, props) {
      const { isLoading, loaded } = state;
      if (isLoading) {
        return <MyLoadingComponent {...state} />;
      }
      return <loaded.MyComponent {...props}/>;
    }
  });
  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // success
});

test('render element without loading prop', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => ({ MyComponent })),
    render: <CodeSplitRenderer />
  });
  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // success
});

test('loadable map success', async () => {
  let LoadableMyComponent = Loadable.Map({
    loader: {
      a: createLoader(200, () => ({ MyComponent })),
      b: createLoader(400, () => ({ MyComponent })),
    },
    loading: MyLoadingComponent,
    render(loaded, props) {
      return (
        <div>
          <loaded.a.MyComponent {...props}/>
          <loaded.b.MyComponent {...props}/>
        </div>
      );
    }
  });

  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // success
});

test('loadable map success without loading prop', async () => {
    let LoadableMyComponent = Loadable.Map({
      loader: {
        a: createLoader(200, () => ({ MyComponent })),
        b: createLoader(400, () => ({ MyComponent })),
      },
      render(state, props) {
        const { loaded } = state;
        return whenLoaded(state, () => (
          <div>
            <loaded.a.MyComponent {...props}/>
            <loaded.b.MyComponent {...props}/>
          </div>
        ));
      }
    });

    let component = renderer.create(<LoadableMyComponent prop="baz" />);
    expect(component.toJSON()).toMatchSnapshot(); // initial
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // loading
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // success
});

test('loadable map element success without loading prop', async () => {
  let LoadableMyComponent = Loadable.Map({
    loader: {
        a: createLoader(200, () => ({ MyComponent })),
        b: createLoader(400, () => ({ MyComponent })),
    },
    render: <CodeSplitMapRenderer />
  });

  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // success
});

test('loadable map error', async () => {
  let LoadableMyComponent = Loadable.Map({
    loader: {
      a: createLoader(200, () => ({ MyComponent })),
      b: createLoader(400, null, new Error('test error')),
    },
    loading: MyLoadingComponent,
    render(loaded, props) {
      return (
        <div>
          <loaded.a.MyComponent {...props}/>
          <loaded.b.MyComponent {...props}/>
        </div>
      );
    }
  });

  let component = renderer.create(<LoadableMyComponent prop="baz" />);
  expect(component.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component.toJSON()).toMatchSnapshot(); // error
});

test('loadable map error without loading prop', async () => {
    let LoadableMyComponent = Loadable.Map({
      loader: {
        a: createLoader(200, () => ({ MyComponent })),
        b: createLoader(400, null, new Error('test error')),
      },
      render(state, props) {
        const { loaded } = state;
        return whenLoaded(state, () => (
          <div>
            <loaded.a.MyComponent {...props}/>
            <loaded.b.MyComponent {...props}/>
          </div>
        ));
      }
    });

    let component = renderer.create(<LoadableMyComponent prop="baz" />);
    expect(component.toJSON()).toMatchSnapshot(); // initial
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // loading
    await waitFor(200);
    expect(component.toJSON()).toMatchSnapshot(); // error
});

describe('preloadReady', () => {
  beforeEach(() => {
    global.__webpack_modules__ = { 1: true, 2: true };
  });

  afterEach(() => {
    delete global.__webpack_modules__;
  });

  test('undefined', async () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
    });

    await Loadable.preloadReady();

    let component = renderer.create(<LoadableMyComponent prop="baz" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('one', async () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1],
    });

    await Loadable.preloadReady();

    let component = renderer.create(<LoadableMyComponent prop="baz" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('many', async () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1, 2],
    });

    await Loadable.preloadReady();

    let component = renderer.create(<LoadableMyComponent prop="baz" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('missing', async () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(200, () => MyComponent),
      loading: MyLoadingComponent,
      webpack: () => [1, 42],
    });

    await Loadable.preloadReady();

    let component = renderer.create(<LoadableMyComponent prop="baz" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('delay with 0', () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(300, () => MyComponent),
      loading: MyLoadingComponent,
      delay: 0,
      timeout: 200,
    });
  
    let loadingComponent = renderer.create(<LoadableMyComponent prop="foo" />);
  
    expect(loadingComponent.toJSON()).toMatchSnapshot(); // loading
  });

  test('getModules', () => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(300, () => MyComponent),
      modules: ['./myModule']
    });
    expect(LoadableMyComponent.getModules()).toEqual(['./myModule']);
  });

  test('getLoader', () => {
    const componentLoader = createLoader(300, () => MyComponent);
    let LoadableMyComponent = Loadable({
      loader: componentLoader
    });
    expect(LoadableMyComponent.getLoader()).toBe(componentLoader);
  });

  test('preload', (done) => {
    let LoadableMyComponent = Loadable({
      loader: createLoader(200, () => MyComponent)
    });

    let LoadableMapComponent = Loadable({
      loader: {
        myComponent: createLoader(400, () => MyComponent)
      }
    });

    const loaders = [LoadableMyComponent.getLoader(), LoadableMapComponent.getLoader()];
    Loadable.preload(loaders).then((modules) => {
      expect(modules).toHaveLength(2);
      done();
    });
  });
});
