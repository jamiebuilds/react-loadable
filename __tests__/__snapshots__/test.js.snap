// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delay and timeout 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`delay and timeout 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`delay and timeout 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":true,"error":null}
</div>
`;

exports[`delay and timeout 4`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`loadable map element success without loading prop 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map element success without loading prop 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map element success without loading prop 3`] = `
<div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
</div>
`;

exports[`loadable map error 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loadable map error 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loadable map error 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":true,"timedOut":false,"error":{}}
</div>
`;

exports[`loadable map error without loading prop 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map error without loading prop 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map error without loading prop 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":true,"timedOut":false,"error":{},"loaded":{"a":{}}}
</div>
`;

exports[`loadable map success 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loadable map success 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loadable map success 3`] = `
<div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
</div>
`;

exports[`loadable map success without loading prop 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map success without loading prop 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null,"loaded":{}}
</div>
`;

exports[`loadable map success without loading prop 3`] = `
<div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
</div>
`;

exports[`loading error 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loading error 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loading error 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":true,"timedOut":false,"error":{}}
</div>
`;

exports[`loading success 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`loading success 4`] = `
<div>
  MyComponent 
  {"prop":"bar"}
</div>
`;

exports[`preload 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`preload 2`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`preload 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`preloadReady delay with 0 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`preloadReady many 1`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`preloadReady missing 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`preloadReady one 1`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`preloadReady undefined 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`render 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`render 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`render 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`render element 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`render element 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`render element 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`render element without loading prop 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null,"loaded":null}
</div>
`;

exports[`render element without loading prop 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null,"loaded":null}
</div>
`;

exports[`render element without loading prop 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`render without loading prop 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null,"loaded":null}
</div>
`;

exports[`render without loading prop 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null,"loaded":null}
</div>
`;

exports[`render without loading prop 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`server side rendering 1`] = `
<div>
  fixture1
</div>
`;

exports[`server side rendering es6 1`] = `
<div>
  fixture2
</div>
`;
